enterprise_subsidy/apps/transaction/api.py

```diff
@@ -0,0 +1,90 @@
+"""
+import logging
+
+import requests
+from django.utils import timezone
+from openedx_ledger.api import reverse_full_transaction
+from openedx_ledger.models import TransactionStateChoices
+
+from enterprise_subsidy.apps.api_client.enterprise import EnterpriseApiClient
+from enterprise_subsidy.apps.fulfillment.api import GEAGFulfillmentHandler
+from enterprise_subsidy.apps.transaction.utils import generate_transaction_reversal_idempotency_key
+
+from .exceptions import TransactionFulfillmentCancelationException
+
+logger = logging.getLogger(__name__)
+
+
+def cancel_transaction_fulfillment(transaction):
+    """
+    Cancels the edx-platform fulfillment (typically the verified enrollment gets moved
+    to the ``audit`` mode).
+    """
+    if transaction.state != TransactionStateChoices.COMMITTED:
+        logger.info(
+            f"[fulfillment cancelation] {transaction.uuid} is not committed, "
+            "will not cancel fulfillment"
+        )
+        raise TransactionFulfillmentCancelationException(
+            "Transaction is not committed"
+        )
+    if not transaction.fulfillment_identifier:
+        logger.info(
+            f"[fulfillment cancelation] {transaction.uuid} has no fulfillment uuid, "
+            "will not cancel fulfillment"
+        )
+        raise TransactionFulfillmentCancelationException(
+            "Transaction has no associated platform fulfillment identifier"
+        )
+
+    try:
+        EnterpriseApiClient().cancel_fulfillment(transaction.fulfillment_identifier)
+    except requests.exceptions.HTTPError as exc:
+        error_msg = (
+            "Error canceling platform fulfillment "
+            f"{transaction.fulfillment_identifier}: {exc}"
+        )
+        logger.exception("[fulfillment cancelation] %s", error_msg)
+        raise TransactionFulfillmentCancelationException(error_msg) from exc
+
+
+def cancel_transaction_external_fulfillment(transaction):
+    """
+    Cancels all related external GEAG allocations for the given transaction.
+
+    raises:
+      FulfillmentException if the related external references for the transaction
+        are not for a GEAG fulfillment provider.
+    """
+    if transaction.state != TransactionStateChoices.COMMITTED:
+        logger.info(
+            f"[fulfillment cancelation] {transaction.uuid} is not committed, "
+            "will not cancel fulfillment"
+        )
+        raise TransactionFulfillmentCancelationException(
+            "Transaction is not committed"
+        )
+
+    for external_reference in transaction.external_reference.all():
+        provider_slug = external_reference.external_fulfillment_provider.slug
+        geag_handler = GEAGFulfillmentHandler()
+        if provider_slug == geag_handler.EXTERNAL_FULFILLMENT_PROVIDER_SLUG:
+            geag_handler.cancel_fulfillment(external_reference)
+        else:
+            raise TransactionFulfillmentCancelationException(f'dont know how to cancel {provider_slug}')
```

Comment:
Ignore if you think this is too pedantic, but should we attempt to cancel all external references opportunistically before throwing an exception if one of them failed?

Reply:
Not a bad idea - I wonder if we maybe don't bother raising at all and instead just log a warning when we get a non-GEAG external reference?

Reply:
Given that would be more predictable than what's written, the pros outweigh the cons IMO.  Just make a best effort and log a warning sounds good.